### PR TITLE
LPS-65498 'Unable to encrypt blank password' error on scheduled import from LDAP

### PR DESCRIPTION
**Pull request by drewbrokke**
This is an update for [LPS-65498](https://issues.liferay.com/browse/LPS-65498).
This is an update for [LPS-65531](https://issues.liferay.com/browse/LPS-65531).

/cc @NorbertKocsis

### CODE REVIEW

**Comment by JorgeFerrer** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: modules/apps/foundation/portal-security/portal-security-ldap/src/main/java/com/liferay/portal/security/ldap/internal/exportimport/LDAPUserImporterImpl.java [hunk 1]
```diff
@@ -1565,8 +1565,10 @@ protected String updateUserPassword(
 			}
 		}
 
-		_userLocalService.updatePassword(
-			userId, password, password, passwordReset, true);
+		if ((password != null) && !password.equals(StringPool.BLANK)) {
```
Comment (JorgeFerrer) [inline modules/apps/foundation/portal-security/portal-security-ldap/src/main/java/com/liferay/portal/security/ldap/internal/exportimport/LDAPUserImporterImpl.java]: Wouldn't this prevent setting the password to blank?

**Comment by JorgeFerrer** [comment]
Pull request submitted to https://github.com/brianchandotcom/liferay-portal/pull/39671. <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by JorgeFerrer** [comment]
Hey @NorbertKocsis, 

I've only forwarded the fix for LPS-65531. Please resend the fix to the second based on the answer to my question in that commit.

**Comment by NorbertKocsis** [comment]
Hi,

Yes, it does, but as far as I know we can't set a blank password. The issue is that we are trying to set the password to blank during scheduled import if we don't use the autogenerated or default password option as in that case we don't get a password.

Br,
Norbert

**Comment by JorgeFerrer** [comment]
Thanks for your answer Norbert. See my comments below:

> Yes, it does, but as far as I know we can't set a blank password. 

Why is that, shouldn't the product allow it? 
If it may be a bug that is solved later I would prefer to avoid this solution since it can come up later if the clause is removed to allow for blank passwords.

> The issue is that we are trying to set the password to blank during scheduled import if we don't use the autogenerated or default password option as in that case we don't get a password.

It seems then that we need to allow for an option that means "no password specified". We could use the value "null" or maybe a special string that represents "leave password as it is".

Additionally, looking at the actual issue it seems that we are doing some encryption which fails for an empty password. Should that encryption be skipped in that case in addition to the other fix (to avoid just covering it)?

**Comment by NorbertKocsis** [comment]
Hi,

Thanks for sharing your concerns.

> Why is that, shouldn't the product allow it? 
> If it may be a bug that is solved later I would prefer to avoid this solution since it can come up later if the clause is removed to allow for blank passwords.
I don't know we should allow it, but our Login portlet currently doesn't allow trying to log in with the password field empty, so it seems we assume the password can't be blank.
Also the admin can't set the password to blank as to form is not sent with the password fields empty.

Should we ask PM about this?

> It seems then that we need to allow for an option that means "no password specified". We could use the value "null" or maybe a special string that represents "leave password as it is".
It could be done, but the current implementation allows admins to change the password setting (like switch from default password to autogenerated) and these changes would take effect on the next import job. If we do this I think we would lose this flexibility.

> Additionally, looking at the actual issue it seems that we are doing some encryption which fails for an empty password. Should that encryption be skipped in that case in addition to the other fix (to avoid just covering it)?
Well I guess this part depends on whether or not we want to allow blank passwords. If we do then you are right and we definitely have to fix this. But if don't then failing with an exception in case of a blank password should be fine.

And thanks for looking into this! :)

Br,
Norbert

**Comment by JorgeFerrer** [comment]
Hey Norbert,

> Should we ask PM about this?

I'm acting as the product manager for User Administration. I would need to do some research to determine whether we want to allow for blank passwords, but I just don't want to introduce that limitation just to solve an issue that could be solved in a different way.

> It could be done, but the current implementation allows admins to change the password setting (like switch from default password to autogenerated) and these changes would take effect on the next import job. If we do this I think we would lose this flexibility.

Why is that the case?

> Well I guess this part depends on whether or not we want to allow blank passwords. If we do then you are right and we definitely have to fix this. But if don't then failing with an exception in case of a blank password should be fine.

As mentioned, I'm not sure yet whether we want to allow them or not, but I don't want to put barriers to doing it if we decide to. Richard Stallman, for example, is known to refuse using any system that does not allow a blank password since he always uses that one.

**Pull request closed without merging**